core/src/main/java/feign/SynchronousMethodHandler.java

```diff
@@ -126,7 +128,7 @@ Object executeAndDecode(RequestTemplate template, Options options) throws Throwa
       if (logLevel != Logger.Level.NONE) {
         logger.logIOException(metadata.configKey(), logLevel, e, elapsedTime(start));
       }
-      throw errorExecuting(request, e);
+      throw errorDecoder.decode(request, e);
```

Comment:
@velo - thanks for taking a look at this. I tried two different approaches because I was also not sure which one would align better with the current usage of the library. See my comment on the issue: https://github.com/OpenFeign/feign/issues/1487#issuecomment-949446474

Do you have any alternative approach in your mind I could experiment with?

Reply:
Honestly, hate current exception creation approach feign is using.

It's confusing, hard to maintain, and even harder to extend/override.

I would be very keen on a major redo for it, I just don't have a good idea on how to do it.

For sure this would trigger a major version release.


regarding your question
https://github.com/gitaroktato/feign/commit/868e2a150f0911bb074d9071a88fadb210c2e44a

this seems to be the best approach.  Just catch the exception on ApacheHttpClient.

Reply:
OK, I reverted the PR to the previous commit as a baseline. 

My only problem with this approach was that the root cause is in the exception's stack trace (see the sample below). I was also thinking about either parsing the `ProtocolException` or adding it as a parameter of the `HttpResponse`. What do you think, how we can make this better?

```java
Caused by: org.apache.http.ProtocolException: Received redirect response HTTP/1.1 303 Redirection but no location header
	at org.apache.http.impl.client.DefaultRedirectStrategy.getLocationURI(DefaultRedirectStrategy.java:148)
	at org.apache.http.impl.client.DefaultRedirectStrategy.getRedirect(DefaultRedirectStrategy.java:223)
	at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:126)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185)
	... 35 more
```

Reply:
Good point

What if we throw an feign exception on line 84? https://github.com/gitaroktato/feign/commit/868e2a150f0911bb074d9071a88fadb210c2e44a#diff-1c348e0c791a430b6d6355068c13167ec2a86ec95cc87b129307b879cd011d13R84

Reply:
If we throw any `RuntimeException` the `SynchronousMethodHandler` is not handling the exception and `ErrorDecoder.decode` is not being called.

See method body in line 119: https://github.com/OpenFeign/feign/blob/3ea3708cddef0d20c0bf443c18f2364c8d8df062/core/src/main/java/feign/SynchronousMethodHandler.java#L119

Reply:
The point of the error decoder is to take a body from a error response and parse into some meaningful response/exception

When you have a `ClientCrashedException` there is no much point on decoding.... cause if the client thrown an exception mean it was unable to handle server somehow

Reply:
OK, so if I understood you right there's no reason to call error decoder meaning that issue #1524 cannot be fully fixed? All that we should do is to wrap with a `RuntimeException`?